src/scanpy/preprocessing/_normalization.py

```diff
@@ -51,14 +49,60 @@ def nonzero_median(x):
     )
 
 
+@njit
+def _normalize_csr(
+    indptr,
+    indices,
+    data,
+    *,
+    rows,
+    columns,
+    target_sum,
+    exclude_highly_expressed: bool = False,
+    max_fraction: float = 0.05,
+    n_threads: int = 10,
+):
+    counts_per_cell = np.zeros(rows, dtype=data.dtype)
+    for i in numba.prange(rows):
+        count = 0.0
+        for j in range(indptr[i], indptr[i + 1]):
+            count += data[j]
+        counts_per_cell[i] = count
+    if exclude_highly_expressed:
+        counts_per_cols_t = np.zeros((n_threads, columns), dtype=np.int32)
+        counts_per_cols = np.zeros(columns, dtype=np.int32)
+
+        for i in numba.prange(n_threads):
+            for r in range(i, rows, n_threads):
+                for j in range(indptr[r], indptr[r + 1]):
+                    if data[j] > max_fraction * counts_per_cell[r]:
+                        minor_index = indices[j]
+                        counts_per_cols_t[i, minor_index] += 1
+        for c in numba.prange(columns):
+            counts_per_cols[c] = counts_per_cols_t[:, c].sum()
+
+        for i in numba.prange(rows):
+            count = 0.0
+            for j in range(indptr[i], indptr[i + 1]):
+                if counts_per_cols[indices[j]] == 0:
+                    count += data[j]
+            counts_per_cell[i] = count
+
+    if target_sum == 0:
+        target_sum = np.median(counts_per_cell)
+    for i in numba.prange(rows):
+        count = counts_per_cell[i] / target_sum
+        for j in range(indptr[i], indptr[i + 1]):
+            data[j] /= count
+    return counts_per_cell, target_sum, counts_per_cols
```

Comment:
so this part can be in fast-array-utils under the name `elem_mult` @flying-sheep right? This is what I understood from our discussions. Should we first implement `elem_mult` then come back to this issue?

Reply:
Generally I wouldn’t stop scanpy improvements for `fast_array_utils`.

But if part of this can be replaced with a mult helper, we could do that now!